File: src/main/java/dev/cbyrne/pufferfishmodloader/gradle/extension/TargetExtension.java
package dev.cbyrne.pufferfishmodloader.gradle.extension;

import dev.cbyrne.pufferfishmodloader.gradle.PufferfishGradle;
import dev.cbyrne.pufferfishmodloader.gradle.mappings.MappingProvider;
import dev.cbyrne.pufferfishmodloader.gradle.mappings.MappingProviderRegistry;
import dev.cbyrne.pufferfishmodloader.gradle.mappings.YarnMappingProvider;
import groovy.lang.Closure;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TargetExtension implements Serializable {
    private final String version;
    private transient final PufferfishGradle plugin;
    private String clientMainClass = "dev.cbyrne.pufferfishmodloader.launch.PMLClientMain";
    private String serverMainClass = "dev.cbyrne.pufferfishmodloader.launch.PMLServerMain";
    private MappingProvider mappings;
    private File runDir;
    private List<String> accessTransformers = new ArrayList<>();

    public TargetExtension(String version, PufferfishGradle plugin) {
        this.version = version;
        this.plugin = plugin;
        if (YarnMappingProvider.isAvailable(version)) {
            yarn();
        } else {
            mcp();
        }
        runDir = new File(plugin.getProject().getProjectDir(), "run/" + version);
    }

    public void accessTransformer(String... transformers) {
        accessTransformers.addAll(Arrays.asList(transformers));
    }

    public List<String> getAccessTransformers() {
        return accessTransformers;
    }

    public void clientMainClass(String main) {
        clientMainClass = main;
    }

    public void serverMainClass(String main) {
        serverMainClass = main;
    }

    public void setClientMainClass(String clientMainClass) {
        this.clientMainClass = clientMainClass;
    }

    public void setServerMainClass(String serverMainClass) {
        this.serverMainClass = serverMainClass;
    }

    public String getClientMainClass() {
        return clientMainClass;
    }

    public String getServerMainClass() {
        return serverMainClass;
    }

    public void runDir(Object dir) {
        runDir = plugin.getProject().file(dir);
    }

    public void setRunDir(Object dir) {
        runDir(dir);
    }

    public File getRunDir() {
        return runDir;
    }

    public void mcp() {
        mappings("mcp");
    }

    public void mcp(Closure<?> closure) {
        mappings("mcp", closure);
    }

    public void yarn() {
        mappings("yarn");
    }

    public void yarn(Closure<?> closure) {
        mappings("yarn", closure);
    }

    public void mappings(String name) {
        MappingProvider provider = MappingProviderRegistry.getMappingProvider(name);
        provider.initialize(plugin, version);
        mappings = provider;
    }

    public void mappings(String name, Closure<?> closure) {
        MappingProvider provider = MappingProviderRegistry.getMappingProvider(name);
        provider.initialize(plugin, version);
        plugin.getProject().configure(provider, closure);
        mappings = provider;
    }

    public MappingProvider getMappings() {
        return mappings;
    }

    public String getVersion() {
        return version;
    }
}

File: src/main/kotlin/me/dreamhopping/pml/gradle/targets/TargetExtension.kt
package me.dreamhopping.pml.gradle.targets

import groovy.lang.Closure
import me.dreamhopping.pml.gradle.PGExtension
import me.dreamhopping.pml.gradle.mappings.CustomMappingProvider
import me.dreamhopping.pml.gradle.mappings.MappingProvider
import me.dreamhopping.pml.gradle.mappings.YarnMappingProvider
import me.dreamhopping.pml.gradle.mappings.YarnMappingProvider.Companion.isYarnAvailable
import org.gradle.api.Project
import org.gradle.api.plugins.JavaPluginConvention
import java.io.File

@Suppress("UNUSED")
class TargetExtension(val minecraft: PGExtension, val project: Project, val version: String, addDefaultMappings: Boolean = true) {
    val mappingProviders = arrayListOf<MappingProvider>()
    var mappingVersion = version
    var accessTransformers = hashSetOf<String>()
    var clientMainClass = "me.dreamhopping.pml.launch.PMLClientMain"
    var serverMainClass = "me.dreamhopping.pml.launch.PMLServerMain"
    var runDir = File(project.projectDir, "run/$version")
    var sourceSetName = "mc$version"
        set(v) {
            val old = field
            field = v
            TargetConfig.onSourceSetNameChange(this, old)
        }

    init {
        project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets.maybeCreate(sourceSetName)
        TargetConfig.setupTarget(this)

        if (addDefaultMappings) {
            if (project.isYarnAvailable(version)) {
                yarn()
            }
        }
    }

    fun clientMainClass(name: String) {
        clientMainClass = name
    }

    fun serverMainClass(name: String) {
        serverMainClass = name
    }

    fun sourceSetName(name: String) {
        sourceSetName = name
    }

    fun mappingVersion(version: String) {
        mappingVersion = version
    }

    @JvmOverloads
    fun yarn(version: String? = null) {
        mappingProviders.add(YarnMappingProvider(version) { onIdChange() })
        onIdChange()
    }

    private fun onIdChange() {
        TargetConfig.onIdChange(project, version, mappingProviders.joinToString("-") { it.id })
    }

    fun mappings(closure: Closure<*>) {
        val provider = CustomMappingProvider(project)
        project.configure(provider, closure)
        mappingProviders.add(provider)
        onIdChange()
    }
}
